### ISSUE TITLE
Feature Request: MemPalace Lite — Zero-Inference Fallback Mode

### ISSUE THREAD

**Issue by duskfallcrew** [milla-jovovich/mempalace#496]
## 💡 Feature Request: MemPalace Lite — Zero-Inference Fallback Mode

**TL;DR**: Add a `--lite` mode that **skips ONNX entirely** — using pure Python regex/keyword matching + TF-IDF scoring — outputting human-readable YAML/SQLITE/plaintext. Optional one-time `init --seed` pulls public, privacy-safe templates. Keeps MemPalace accessible for users on constrained hardware without compromising the local-first, zero-cloud philosophy.

---

### 🙏 Context & Gratitude
First, thank you for building MemPalace — it's a genuinely thoughtful approach to local-first, privacy-respecting memory management. 

This suggestion comes from a real-world block: I tried running `mempalace mine` on **[Ktiseos-Nyx-Trainer](https://github.com/Ktiseos-Nyx/Ktiseos-Nyx-Trainer/)** — a **LoRA training UI built on Kohya SS sd-scripts with a Next.js + FastAPI frontend** — on my **iMac 2019 (AMD Radeon Pro 580, 8GB VRAM) running Windows via Bootcamp**.

**What happened**:
- The `mine` command began scanning the **Kohya SS sd-scripts Python codebase** (the actual training logic) — thousands of lines of complex, nested Python
- After ~45 minutes, the ONNX pipeline hit a hard memory wall:  
  > 775 Files it was trying to validly mine, and after 30 minutes of trying just a singular file the 111th file just wasn't "Mining". 

This isn't a critique of MemPalace — it's about **inclusive design**. Large, complex repos aren't edge cases: they're the norm for ML tooling, training frameworks, and legacy codebases. If the default pipeline requires hardware many users don't have, we risk locking out:
- Neurodivergent folks who rely on structured memory aids but can't upgrade hardware
- Users on disability benefits or limited budgets
- Developers in regions with limited access to high-end GPUs
- Anyone maintaining legacy machines that still "work fine" for everything else

*(Side note: If I tried this on my older PyQt6 project that I rewrote in Node/Next? I shudder to think. 😅)*

---

### 🔹 Core Proposal: `--lite` = Zero ONNX, Pure Python

| Feature | Full Mode | Lite Mode |
|---------|-----------|-----------|
| **Embedding** | Default ONNX model (~420MB) | **None** — regex/keyword + TF-IDF only *(pure Python, zero ML)* |
| **Storage** | ChromaDB vector store | YAML / plaintext / SQLITE (human-readable, git-diffable) |
| **CLI Discovery** | `mempalace mine` | `mempalace mine --lite` OR config key `mode: lite` |
| **Hardware Floor** | GPU / Apple Silicon recommended | Runs on any Python 3.8+ machine |
| **Tradeoff** | Semantic richness, vector search | Keyword-based ranking; still usable context for LLMs |

**Clarification**: Lite mode **does not load any ONNX runtime, not even a small model**. It's a deliberate, explicit opt-out of the embedding pipeline for users whose hardware says "nope". 

**Why this works**:
- Zero inference cost → no OOM errors, no 45-minute waits that end in failure
- Output remains LLM-consumable: structured YAML/SQLITE files any agent can read
- Users explicitly opt into the tradeoff; full mode remains default
- **Applies to ANY workflow**: ML training frameworks, genealogy tools, creative projects, research repos — if it has code or text, Lite mode can index it

---

### 🔹 Real-World Example: Why Lite Mode Matters

Take **[Ktiseos-Nyx-Trainer](https://github.com/Ktiseos-Nyx/Ktiseos-Nyx-Trainer/)**:
- **Stack**: Next.js + FastAPI frontend, Kohya SS sd-scripts Python backend
- **Scope**: 132+ training parameters, support for SDXL/SD1.5/Flux/SD3, multiple LoRA types (Standard, LoCon, LoHa, LoKr, DoRA), Civitai/HF model browser, dataset tagging, caption editing, etc.
- **Mining attempt**: Scanned through the **sd-scripts Python training code** (thousands of lines of complex logic) → ONNX OOM on 8GB AMD GPU after ~45 minutes
- **With Lite mode**: 
  - Pure Python regex/TF-IDF quickly flags key patterns: `train_network.py`, `config_presets`, `lora_injection`, `caption_parser`, `api_endpoints`
  - Outputs `ktiseos_nyx_lite.yaml` with ranked sections + file paths for an LLM to consume
  - User gets *usable context* in minutes, not hours — even if it's keyword-based, not semantic

This isn't about replacing full mode — it's about **having a fallback that still delivers value** when hardware says "nope". Also my training service app works with the back-end being a vendor version of sd-scripts, which is a "battle-tested" as LLMs like to call it, community training code. I've used it in notebooks for three years, it's roughly all I know in terms of how to do open source training.

---

### 🔹 "Lite Seed" Template System (Privacy-Respecting)

Optional one-time scaffold download:
```bash
mempalace init --seed  # pulls from curated public repo, then fully local
```

**What's in a seed**:
- Pre-built "wing/room/hall" templates for common workflows:
  - 🧠 ML training: Python, Tensorflow, Diffusers, parameter groups
  - 💻 Web dev: Next.js app router patterns, FastAPI route scaffolds
  - 🔬 Research: genealogy data formats, citation templates
  - ✍️ Creative: story beat templates, character sheet structures
  - 🧠 Neurodivergent Skills: instead of downloading custom skills for each individual ND need up front, common patterns could be built in. 
- Community-contributed, fact-checked examples in Markdown/YAML/TOML
- **No private data** — only public, anonymized scaffolding

**Privacy alignment**:
- GitHub acts only as a **static template CDN** (like shadcn/ui)
- No ongoing sync, no telemetry, no cloud dependency after init
- User's actual memories stay 100% local, never leave the machine
- Templates can be offered to then be added either as a registry (Like Shadcn) OR to add as a feature request to Mempalace in general. 

---

### 🔹 Implementation Sketch (Low-Lift Hooks)

If this resonates, here are concrete starting points:
1. Add a `--mode` CLI arg: `full` (default) | `lite`
2. For `lite`: 
   - Skip ONNX loader entirely
   - _Use `re` + `sklearn.feature_extraction.text.TfidfVectorizer` (optional dep, or pure Python fallback) [Please note, I am not sure SciKit learn is going to actually help if one doesnt' have a decent GPU or CPU, this was something Qwen suggested, but i'm not sure about it?]_
   - Output adapter: `--format yaml` | `sqlite` | `txt` instead of ChromaDB
3. Seed system: a `templates/` folder in a public repo, fetched once via GitHub raw CDN

**Zero new heavy dependencies** — Lite mode should work with just `pip install mempalace[lite]` or even vendored stdlib where possible.

---

### 🔹 Why This Matters Beyond "Potato PCs"

- ♿ **Accessibility**: Hardware constraints shouldn't block access to useful tooling
- 🌍 **Equity**: Users in regions with limited hardware access deserve the same core functionality
- 🧠 **Neurodivergent inclusion**: Lower cognitive + technical barrier to entry = more sustainable workflows
- 🔐 **Privacy-first preserved**: Lite mode doesn't compromise the zero-cloud ethos — it extends it to more machines
- 🔄 **Workflow-agnostic**: Works for ML training code, web apps, research notes, creative writing, genealogy databases — any text-based knowledge base

---

### 🤝 Willing to Help
Happy to:
- Test Lite mode builds on my AMD 8GB VRAM Bootcamp setup
- Contribute example seed templates (e.g., Kohya SS training patterns, Next.js/FastAPI project structures, genealogy research workflows)
- Help document the tradeoffs for users choosing `--lite`
- Provide real-world feedback from the Ktiseos-Nyx codebase as a test case

Either way, thank you again for the project — it's already inspiring great workflows, and I'd love to see it reach even more users. ✨

---

### ✅ Why this framing works for maintainers:
- **Concrete, accurate example**: Ktiseos-Nyx-Trainer = LoRA UI on Kohya SS + Next.js/FastAPI; mining failed on sd-scripts Python code
- **Zero-ONNX clarity**: Explicitly states Lite mode skips ML entirely (no ambiguity)
- **Implementation hooks**: CLI flags, pure-Python libs, output formats — actionable
- **Privacy-aligned**: GitHub = one-time template CDN, not a sync backend
- **Accessibility-framed**: Not just "make it faster", but "include more users"
- **Workflow-agnostic**: Emphasizes this helps ANY project type, especially large/complex ones

P.s. I used Qwen 3.6 i think it was to construct the framing in a way that was more able to be read easier for developers and others.  I pitched this also towards ONE other neurodivergent who is coincidentally my partner in life, and business and I was just teaching them the ropes of my trainer app today, and a lite version would even service newer ai assistance users who have lower end PCs! <3

**Comment by web3guru888** [comment]
This is a well-framed proposal and the real-world failure case makes it concrete.

A few thoughts from running MemPalace at scale on similar constraints:

**The TF-IDF fallback is genuinely useful, not a consolation prize.** For code repositories especially — where the vocabulary is highly consistent and term frequency is meaningful — TF-IDF retrieval often beats embedding-based search for exact symbol lookups (function names, class names, config keys). The semantic richness tradeoff is real but it's smaller than it sounds for developer-facing use cases.

**scikit-learn's TfidfVectorizer doesn't need a GPU.** It's pure NumPy/SciPy and runs fine on CPU. The concern in your issue about whether sklearn helps on low-end hardware is valid for *training* ML models, but TF-IDF is just matrix math — 8GB of CPU RAM is more than enough for a typical codebase index.

**The embedding model size is already configurable via ChromaDB defaults.** You can swap in `all-MiniLM-L6-v2` (~80MB, sentence-transformers) instead of the default larger model without any code changes — just update the ChromaDB embedding function. A `--model lite` flag that hard-codes this smaller model and sets ONNX to CPU-only would address the OOM case without requiring a fully separate lite pipeline.

**Two things I'd add to the lite mode proposal:**

1. **Chunked file processing.** The 775-file queue hitting OOM after file #111 suggests the issue isn't just model size but memory accumulation across files. A lite mode should also process files one-at-a-time with explicit garbage collection between files (`gc.collect()` after each), not batch them in a queue.

2. **Progress persistence.** Mining a large codebase in lite mode should checkpoint progress to disk so an interrupted run resumes from where it stopped, not from file #1. This is especially important on constrained hardware where OOM kills are more likely.

**Seed templates** are a great idea independently of lite mode. A `mempalace init --seed <workflow-type>` that pulls public YAML scaffolding for common project structures would reduce the blank-palace friction for new users significantly. The privacy alignment (one-time pull, fully local after) is correct.

The AMD 580 + Bootcamp setup is a real constraint worth designing for. Happy to help test a lite mode implementation against our own large-palace use case if this moves toward a PR.

**Comment by duskfallcrew** [comment]
Cool, I didn't realize that Scikit was NumPY based, I should have honestly recognized that through using stable diffusion adjacent tools.

I welcome ANY additions to my pitch, honestly this was like a 5 minute pitch after I finished cooking dinner.  

I originally had suggested a long term semi "compromise" where users could consistently use MCP servers for github as a fallback, which in turn though would say "ItS NOT AS PRIVATE" - so that was why we came up with the seeded idea. 

The same concept would've flown for the lesser idea of github reading directly, so either way we would've still had seeded templates.

I'm actually not naturally a developer despite the fact that i'm consistently "blowing" braincells on coding projects that nobody asks for. I was a smidge gutted  I didn't read the readme well, and that was on me, and same with Claude - we didn't read it entirely ,and i saw the ML stack go through the PIP install and was like "Help, my CPU has a death wish on me." 

I didn't get a FULL OOM: But it was going to likely lock my computer up, my CPU is fairly decent despite it being a potato "tin can" as I call it. But the idea that the initial mine is "SLOW" wasn't my issue like stated in the pitch, it was that even AFTER a 40 minute wait the next file wasn't complete. It was 20 minutes or more for a file that seemed to be i think 1000 lines or was it 1000 tokens? 

But the seedbase would allow the idea of regex, or even like Qwen and you stated: Scikit+NumPY (which i'm slightly familiar with when I was doing the pyqt6 version of Dataset-Tools, now re-coded in Next/Node) which would then service things like privacy-first templates based on user provided codebases.

**Comment by duskfallcrew** [comment]
Ah adding one thing that i just forgot when I was mentioning this on threads:

Even the lite mode itself could then be used for the full mode, allowing an extra buffer to the normal mining service for the AI assistance. Someone just starting a project after mining could still use the publically constructed seedbank to aid in memory collection. 

So often we're using "SKILLS" for things that don't need SKILLS, they need a "REFERENCE BASE" - Skills are incredibly useful in Agentic AI.  Yet, do we need a skill for every codebase? No, we need a memory bank, just like mempalace does for collecting and mining information in the full version, the lite version could ascertain the information from the seedbank as an optional addition. 

Like the initial "ROOMS" of the main mempalace structure, the seedbank files could contain references to existing or "NEW ROOMS". Like stated in my initial pitch, this could mean that people don't have to contain millions of skills for an agent in how to handle neurodiversity, references are just like that with humans, why are we throwing program logic all the time as a "REFERENCE" rather than treating it like a human thing? Not that we can't use programming logic to solve the ACTUAL code, but mempalace feels like this: A Library of your code. The seedbank feels like: When you take your homework with you to an actual library and upgrade your "knowledge base" and "research new ways of doing things.". 

I think this is an incredibly wonderful way of doing it because so often Claude and other programs tend to forget how certain functions work, just like humans - because humans invented LLM's, and we trained them. 

So in ways to usher a more "ethical" (it's AI, is it ever fully ethical?) future we have a new library that can "GUIDE" without a full finetune or re-train!

**Comment by web3guru888** [comment]
The tiered approach you're describing — lite output as a pre-index for the full ONNX pass — is actually quite clever. Rather than lite being a permanent fallback, it becomes a first-pass triage layer: TF-IDF extracts candidates quickly, ONNX re-ranks only the top N. That would help on constrained hardware because you're spending the expensive embedding budget on a filtered set rather than the full corpus.

The 'bootstrap from lite, promote to full' lifecycle you're sketching maps roughly to what we do with staged discovery: a cheaper scan identifies candidate drawers, the expensive semantic pass validates them. The difference is we do it at query time; your proposal puts it at mine time. Both approaches reduce unnecessary compute.

One thing to flag: the scikit-learn TF-IDF vectorizer still pulls in numpy/scipy, which isn't trivial on some constrained targets. A pure-Python implementation (no scipy) would keep the dependency surface smaller — the vocabulary size in typical MemPalace usage is small enough that a pure dict-based TF-IDF would be fast enough for lite mode without the extra weight.

Happy to contribute a proof-of-concept lite miner if this gets traction as a proper issue.

**Comment by duskfallcrew** [comment]
I have no clue how to develop anything so if anything i'd be the tester-monkey when something like this would come about.  I'm happy to do problem solving ideation and even testing, just that I wouldn't know where to start with developing it.

**Comment by web3guru888** [comment]
Testing is a real contribution — especially for a feature like this where the failure modes are subtle. Knowing whether the lite-mode output degrades gracefully vs silently-wrong across different context sizes is exactly the kind of thing that's hard to validate without real workloads.

If this gets built, flagging patterns where the lite pre-index steers the full pass toward the wrong drawer would be high-value bug data. The interesting failure case would be: lite-mode classifies something into Hall A, full ONNX pass should have gone to Hall B, but the pre-index biases it wrong. That kind of systematic drift is what would make or break the tiered approach in practice.

**Comment by rebroad** [comment]
https://github.com/rebroad/mempalace

I've currently set up a lite version, which I'm using on a 1GB Raspberry Pi3 (running Debian 12). It was largely "vibe coded" though, but I'm using it and it seems to work.

**Comment by duskfallcrew** [comment]
> https://github.com/rebroad/mempalace
> 
> I've currently set up a lite version, which I'm using on a 1GB Raspberry Pi3 (running Debian 12). It was largely "vibe coded" though, but I'm using it and it seems to work.

It's a start, there was a lot of context towards making sort of pre-indexable information about certain types of code i think in my feature request - i know that hte idea is to keep it centralized and local, but at the same time additive features like that even if it's another repo that's pre-checked, might be something interesting..

Something to lessen load.

See how yours rolls for a while, i'm good for now as anyhting even on debian MAY SITLL NOT WORK on my stupid potato XD  IN MY HUMBLE OPINION my imac running windows is probably slower than a raspberry pi lmao
